File: libs/power-log-parser/src/lib/parsers/battlegrounds-opponent-revealed-parser.ts
import { CardIds, CardType, GameTag, GameType, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEvent, GameEventProvider } from '../game-event';
import { FullEntity, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

const EXCLUDED_HERO_CREATOR_DBFIDS: number[] = [63600];

export class BattlegroundsOpponentRevealedParser implements ActionParser {
	readonly ParserName = 'BattlegroundsOpponentRevealedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.FullEntity &&
			(node.Object as FullEntity).GetTag(GameTag.ZONE) === (Zone.SETASIDE as number) &&
			(node.Object as FullEntity).GetTag(GameTag.CARDTYPE) === (CardType.HERO as number)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const fullEntity = node.Object as FullEntity;
		const cardId = fullEntity.CardId;
		if (
			cardId === CardIds.BaconphheroHeroic ||
			cardId === CardIds.TagtransferplayerenchantDntEnchantment_Bacon_TagTransferPlayerE
		) {
			return null;
		}

		const result: GameEventProvider[] = [];
		result.push(
			GameEventProvider.Create(
				fullEntity.TimeStamp,
				'BATTLEGROUNDS_OPPONENT_REVEALED',
				() => this.buildGameEvent(node),
				false,
				node,
			),
		);
		if (
			this.GameState.NextBgsOpponentPlayerId > 0 &&
			fullEntity.GetTag(GameTag.PLAYER_ID) === this.GameState.NextBgsOpponentPlayerId &&
			!EXCLUDED_HERO_CREATOR_DBFIDS.includes(fullEntity.GetTag(GameTag.CREATOR_DBID))
		) {
			result.push(
				GameEventProvider.Create(
					fullEntity.TimeStamp,
					'BATTLEGROUNDS_NEXT_OPPONENT',
					() => {
						if (!this.StateFacade.IsBattlegrounds()) {
							return null;
						}
						return {
							Type: 'BATTLEGROUNDS_NEXT_OPPONENT',
							Value: {
								CardId: cardId,
								OpponentPlayerId: fullEntity.GetTag(GameTag.PLAYER_ID),
								LeaderboardPlace: fullEntity.GetLeaderboardPosition(
									this.StateFacade.GetMetaData().GameType as GameType,
								),
							},
						};
					},
					true,
					node,
				),
			);
			this.GameState.BgsHasSentNextOpponent = true;
			this.GameState.NextBgsOpponentPlayerId = -1;
		}
		return result;
	}

	private buildGameEvent(node: Node): GameEvent | null {
		const fullEntity = node.Object as FullEntity;
		const cardId = fullEntity.CardId;
		if (!this.StateFacade.IsBattlegrounds()) {
			return null;
		}

		if (this.StateFacade.OpponentPlayer?.PlayerId !== fullEntity.GetEffectiveController()) {
			return null;
		}

		if (fullEntity.GetTag(GameTag.BACON_IS_KEL_THUZAD) === 1) {
			return null;
		}

		return {
			Type: 'BATTLEGROUNDS_OPPONENT_REVEALED',
			Value: {
				CardId: cardId,
				PlayerId: fullEntity.GetTag(GameTag.PLAYER_ID),
				LeaderboardPlace: fullEntity.GetLeaderboardPosition(
					this.StateFacade.GetMetaData().GameType as GameType,
				),
				Health: fullEntity.GetTag(GameTag.HEALTH),
				Armor: fullEntity.GetTag(GameTag.ARMOR),
			},
		};
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsOpponentRevealedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;
using System;
using HearthstoneReplays.Events.Parsers.Utils;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsOpponentRevealedParser : ActionParser
    {
        private static IList<int> EXCLUDED_HERO_CREATOR_DBFIDS = new List<int>()
        {
            63600 // TB_BaconShop_HP_081
        };

        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsOpponentRevealedParser(ParserState ParserState, StateFacade stateFacade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = stateFacade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(FullEntity)
                && (node.Object as FullEntity).GetTag(GameTag.ZONE) == (int)Zone.SETASIDE 
                && (node.Object as FullEntity).GetTag(GameTag.CARDTYPE) == (int)CardType.HERO;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var cardId = fullEntity.CardId;
            if (BgsUtils.IsBaconEnchantment(cardId))
            {
                return null;
            }

            var result = new List<GameEventProvider>();
            result.Add(GameEventProvider.Create(
                fullEntity.TimeStamp,
                "BATTLEGROUNDS_OPPONENT_REVEALED",
                () => BuildGameEvent(node),
                false,
                node)
            );
            if (GameState.NextBgsOpponentPlayerId > 0 
                && fullEntity.GetTag(GameTag.PLAYER_ID) == GameState.NextBgsOpponentPlayerId
                && !EXCLUDED_HERO_CREATOR_DBFIDS.Contains(fullEntity.GetTag(GameTag.CREATOR_DBID)))
            {
                result.Add(GameEventProvider.Create(
                        fullEntity.TimeStamp,
                        "BATTLEGROUNDS_NEXT_OPPONENT",
                        () =>
                        {
                            if (!StateFacade.IsBattlegrounds())
                            {
                                return null;
                            }
                            return new GameEvent
                            {
                                Type = "BATTLEGROUNDS_NEXT_OPPONENT",
                                Value = new
                                {
                                    CardId = cardId,
                                    OpponentPlayerId = fullEntity.GetTag(GameTag.PLAYER_ID),
                                    LeaderboardPlace = fullEntity.GetLeaderboardPosition((GameType)StateFacade.GetMetaData().GameType),
                                }
                            };
                        },
                        true,
                        node));
                GameState.BgsHasSentNextOpponent = true;
                GameState.NextBgsOpponentPlayerId = -1;
            }
            return result;
        }

        private GameEvent BuildGameEvent(Node node)
        {
            var fullEntity = node.Object as FullEntity;
            var cardId = fullEntity.CardId;
            if (!StateFacade.IsBattlegrounds())
            {
                return null;
            }

            if (StateFacade.OpponentPlayer?.PlayerId != fullEntity.GetEffectiveController())
            {
                return null;
            }

            // Having the ghost as a new opponent doesn't tell us anything useful, as there is no
            // link between the ghost and the dead player
            if (fullEntity.GetTag(GameTag.BACON_IS_KEL_THUZAD) == 1)
            {
                return null;
            }

            return new GameEvent
            {
                Type = "BATTLEGROUNDS_OPPONENT_REVEALED",
                Value = new
                {
                    CardId = cardId,
                    PlayerId = fullEntity.GetTag(GameTag.PLAYER_ID),
                    LeaderboardPlace = fullEntity.GetLeaderboardPosition((GameType)StateFacade.GetMetaData().GameType),
                    Health = fullEntity.GetTag(GameTag.HEALTH),
                    Armor = fullEntity.GetTag(GameTag.ARMOR),
                }
            };
        }
    }
}
